### PR TITLE
safeweb: replace gorilla with Sec-Fetch-Site check

### PR DESCRIPTION
**Pull request by patrickod**
Require that all non-(GET|OPTIONS|HEAD) requests to the browser mux specify Sec-Fetch-Site=same-origin to prohibit cross-origin requests.

Updates tailscale/corp#25340

### CODE REVIEW

File: safeweb/http.go [hunk 1]
```diff
@@ -203,6 +181,10 @@ type Config struct {
 	// Do not use the Handler field of http.Server, as it will be ignored.
 	// Instead, set your handlers using APIMux and BrowserMux.
 	HTTPServer *http.Server
+
+	// ServeHSTS specifies whether to serve the Strict-Transport-Security header
+	// in response to requests made to the BrowserMux.
+	ServeHSTS bool
```
Comment (awly) [inline safeweb/http.go:187]: could we make this opt-out instead of opt-in?

File: safeweb/http.go [hunk 2]
```diff
@@ -376,10 +337,15 @@ func (s *Server) serveBrowser(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Security-Policy", s.csp)
 	w.Header().Set("X-Content-Type-Options", "nosniff")
 	w.Header().Set("Referer-Policy", "same-origin")
-	if s.SecureContext {
+	if s.ServeHSTS {
 		w.Header().Set("Strict-Transport-Security", cmp.Or(s.StrictTransportSecurityOptions, DefaultStrictTransportSecurityOptions))
 	}
-	s.csrfProtect(s.BrowserMux).ServeHTTP(w, r)
+	if !slices.Contains([]string{"GET", "HEAD", "OPTIONS"}, r.Method) && r.Header.Get("Sec-Fetch-Site") != "same-origin" {
```
Comment (awly) [inline safeweb/http.go:343]: should there be an option to allow `same-site`?

File: safeweb/http.go [hunk 3]
```diff
@@ -376,10 +337,15 @@ func (s *Server) serveBrowser(w http.ResponseWriter, r *http.Request) {
...
+		// deny non-GET/HEAD/OPTIONS requests that are not same-origin
+		http.Error(w, "forbidden cross-origin request", http.StatusForbidden)
```
Comment (awly) [inline safeweb/http.go:345]: let's return a different error if `Sec-Fetch-Site` is not sent by the client instructing them to update their browser

File: safeweb/http_test.go [hunk 4]
```diff
@@ -208,23 +218,8 @@ func TestCSRFProtection(t *testing.T) {
...
-				s.h.Handler.ServeHTTP(w, get)
-				resp := w.Result()
-
-				// pass the token & cookie in our subsequent test request
-				req.Header.Set("X-CSRF-Token", token)
-				for _, c := range resp.Cookies() {
-					req.AddCookie(c)
-				}
+			if tt.setSecFetchSite {
+				req.Header.Set("Sec-Fetch-Site", "same-origin")
```
Comment (awly) [inline safeweb/http_test.go:222]: nit: also add a test case with a different non-empty header value

File: safeweb/http.go (hunk 2 above)
Comment (awly) [inline safeweb/http.go:343]: WDYT about adding a metric to track this firing?

Comment (patrickod) [reply safeweb/http.go:343]: I can't think of an internal case where we need it but we can def allow it via configuration :thumbsup:

File: safeweb/http.go [hunk 5]
```diff
@@ -376,10 +343,38 @@ func (s *Server) serveBrowser(w http.ResponseWriter, r *http.Request) {
...
 	}
-	s.csrfProtect(s.BrowserMux).ServeHTTP(w, r)
+
+	// Allow GET, HEAD, and OPTIONS requests to the browser mux without
+	// Sec-Fetch-Site header checks.
+	switch r.Method {
+	case "GET", "HEAD", "OPTIONS":
+		s.BrowserMux.ServeHTTP(w, r)
+		return
+	default:
```
Comment (creachadair) [inline safeweb/http.go:356]: FWIW, the `default` isn't needed here (although it is also harmless).

(This is in contrast with a `select` where a default case has a material effect on the fallthrough behaviour)

**Review by creachadair — COMMENTED** [review]
drive-by

File: safeweb/http.go (hunk 5 above)
Comment (patrickod) [reply safeweb/http.go:356]: an educational drop-in comment is the best type. TIL. thanks!
